### PR TITLE
Auth feature 2

### PR DESCRIPTION
**Pull request by snowzhop**
### JWT authentication

- added new classes for _JWT authentication_: **JwtAuthenticationFilter** and **JwtAuthorizationFilter**;
- added new class which includes JWT properties;
- **SecurityConfiguration** extended for JWT auth;
- added new method for user registration;
- user class field 'login' renamed to 'username'.

### CODE REVIEW

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package ikpi63holding.postclient;
+
+public class JwtProperties {
+    public static final String SECRET = "ikpi63HOLDING";
```
Comment (IReallyHateLogins) [inline Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java]: So secret, it's in our open repository. I think we should put most of this class contents in app proprties so it can be changed without compiling.

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+        Authentication authentication = getUsernamePasswordAuthentication(request);
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+
+        chain.doFilter(request, response);
+    }
+
+    private Authentication getUsernamePasswordAuthentication(HttpServletRequest request) {
+        String token = request.getHeader(JwtProperties.HEADER_STRING)
+                .replace(JwtProperties.TOKEN_PREFIX,"");
```
Comment (IReallyHateLogins) [inline Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java]: Why do we replace headers?

File: Spring-server/src/main/java/ikpi63holding/postclient/ApiController.java [hunk 3]
```diff
@@ -122,4 +124,27 @@ void deleteUser(@PathVariable Long id) {
...
+    User registration(@RequestBody User newUser) {
+        PasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
+        
+        if (this.repository.findByUsername(newUser.getUsername()) != null) {
+            throw new ResponseStatusException(HttpStatus.ACCEPTED, "PAYLOAD");
+        }
+
+        return repository.save(new User(
+            newUser.getUsername(),
+            passwordEncoder.encode(newUser.getPassword()),
```
Comment (IReallyHateLogins) [inline Spring-server/src/main/java/ikpi63holding/postclient/ApiController.java:132]: What happends when password is null? Maybe we should check for it implicitly.

**Review by IReallyHateLogins — DISMISSED** [review]
Seems great, I will say more after some testing.

**Review by Daniel1898 — DISMISSED** [review]
Good job!

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.algorithms.Algorithm;
+
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+// import org.springframework.security.web.AuthenticationEntryPoint;
```
Comment (ValentinBodrov) [inline Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java]: Take these comments off in future comments please.

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java (hunk 1 above)
Comment (ValentinBodrov) [reply Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java]: Agreed

File: Spring-server/src/main/java/ikpi63holding/postclient/LoginViewModel.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    public void setUsername(String username) {
+        this.username = username;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+}
```
Comment (ValentinBodrov) [inline Spring-server/src/main/java/ikpi63holding/postclient/LoginViewModel.java]: Is this file necessary? We have already **User.java**, you could use some fields of them by auth process

**Review by ValentinBodrov — COMMENTED** [review]
I didn't understand some points below

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java (hunk 2 above)
Comment (snowzhop) [reply Spring-server/src/main/java/ikpi63holding/postclient/JwtAuthorizationFilter.java]: Because JWT token looks like "Bearer <payload>". This replacing method deletes useless "Bearer " in header for using token in the future

File: Spring-server/src/main/java/ikpi63holding/postclient/ApiController.java (hunk 3 above)
Comment (snowzhop) [reply Spring-server/src/main/java/ikpi63holding/postclient/ApiController.java:132]: If password is null we'll get NullPointerException but i think we should check it before sending of the registration request

File: Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java (hunk 1 above)
Comment (snowzhop) [reply Spring-server/src/main/java/ikpi63holding/postclient/JwtProperties.java]: Yes, we should hide it

File: Spring-server/src/main/java/ikpi63holding/postclient/LoginViewModel.java (hunk 5 above)
Comment (snowzhop) [reply Spring-server/src/main/java/ikpi63holding/postclient/LoginViewModel.java]: Oh, i forgot about User class.

**Review by ValentinBodrov — APPROVED** [review]
It's allright, there's no things those call any questions
